# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character：角色姓名。
- role：演职员类型，演员或导演。

## 读取数据

导入数据分析所用到的库，并通过Pandas的read_csv函数，将原始数据文件“titles.csv”的内容，解析为DataFrame并赋值给original_titles变量。将原始数据文件“credits.csv”的内容，解析为DataFrame并赋值给original_credits变量

In [1]:
import pandas as pd

In [2]:
original_titles = pd.read_csv('titles.csv')
original_credits = pd.read_csv('credits.csv')

In [3]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [4]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

## 评估和清理数据

在这一部分，我们将对上一部分建立的`original_titles`和`original_credits`DataFrame所包含的数据进行评估和清理

主要从两个方面进行：结构和内容，即整齐度和干净度

数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准；数据的内容性问题包括存在缺失数据，重复数据，不一致数据，无效数据及错误数据等

为了区分经过清理的数据和原始的数据，我们创建新的变量cleaned_titles，让它为original_titles复制出的副本；以及创建新的变量cleaned_credits，让它为original_credits复制出的副本。我们之后的清理步骤都将运用在cleaned_titles和cleaned_credits上

In [5]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### 评估数据整齐度

In [6]:
cleaned_titles.sample(10)

id                                        title   type  \
2586    ts83247                                    Stay Here   SHOW   
1439    ts42089                              Haters Back Off   SHOW   
5486  tm1040688                      Three Songs for Benazir  MOVIE   
4319   tm882606                           Sethum Aayiram Pon  MOVIE   
425     ts63295                              Vientos de agua   SHOW   
4534   tm822272                              Love for Sale 2  MOVIE   
5683   ts316882  Reasonable Doubt: A Tale of Two Kidnappings   SHOW   
665    tm178201           Sherlock Holmes: A Game of Shadows  MOVIE   
3843   tm898262                                        Coven  MOVIE   
2719   tm319310                      CIA: Comrade In America  MOVIE   

                                            description  release_year  \
2586  Designer Genevieve Gorder and real estate expe...          2018   
1439  Delve into the oddball family life of Miranda ...          2016   
5486  The story of Shaista, a young man who—newly ma...          2021   
4319  Meeting after many years, a makeup artist and ...          2019   
425   Vientos de agua is a 2006 cult Argentine- Span...          2006   
4534  Tired of matchmaking by his mother, Ican bring...          2019   
5683  A fender bender that turned into a kidnapping ...          2021   
665   There is a new criminal mastermind at large (P...          2011   
3843  French Basque Country, year 1609. The men of a...          2020   
2719  A young Aji Mathew is a communist who goes the...          2017   

     age_certification  runtime  \
2586             TV-PG       32   
1439             TV-14       29   
5486               NaN       22   
4319                 G      102   
425              TV-MA       73   
4534               NaN       92   
5683             TV-MA       47   
665              PG-13      129   
3843               NaN       92   
2719               NaN      134   

                                                 genres production_countries  \
2586                       ['reality', 'documentation']               ['US']   
1439           ['comedy', 'drama', 'family', 'romance']               ['US']   
5486                                          ['drama']               ['AF']   
4319                                ['drama', 'family']               ['IN']   
425                                           ['drama']         ['AR', 'ES']   
4534                               ['drama', 'romance']               ['ID']   
5683                         ['crime', 'documentation']               ['MX']   
665                     ['action', 'crime', 'thriller']         ['GB', 'US']   
3843  ['drama', 'history', 'thriller', 'horror', 'ac...   ['ES', 'FR', 'AR']   
2719                     ['action', 'romance', 'drama']               ['IN']   

      seasons     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
2586      1.0   tt8773338         7.4      1101.0            2.490       7.000  
1439      2.0   tt5467814         5.9      7274.0            7.102       6.400  
5486      NaN  tt14608922         7.7         NaN            3.333       6.400  
4319      NaN   tt8976418         7.2       349.0            1.974       6.000  
425       1.0   tt0459724         8.8      1051.0            1.410       8.800  
4534      NaN  tt10643938         7.0       558.0            2.626       7.100  
5683      1.0  tt15796900         7.6       195.0            3.497       7.000  
665       NaN   tt1515091         7.4    449965.0           51.119       7.100  
3843      NaN   tt9916362         6.4      4754.0            7.981       6.752  
2719      NaN   tt5925772         6.4      3088.0            2.548       6.000

从随机抽样的10条数据来看，变量`genres`和`production_countries`存在多个值，应当进行拆分。

先来获取任意一个`genres`值看看

In [7]:
cleaned_titles['genres'][1]

"['drama', 'crime']"

可以看到，`genres`值的数据类型是字符串，我们要先转换成列表，才能进行拆分

In [8]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(lambda x: eval(x))
cleaned_titles['genres'][1]

['drama', 'crime']

转换成列表后，就可以用DataFrame的explode方法，把那个列的列表值拆分成单独的行

In [9]:
cleaned_titles = cleaned_titles.explode('genres')
cleaned_titles.sample(10)

id                               title   type  \
1729    ts77398                       Lost in Space   SHOW   
413     ts17028                       Dinosaur King   SHOW   
3730    ts88841              The School Nurse Files   SHOW   
2844   tm325610           Chris D'Elia: Man on Fire  MOVIE   
5101  tm1087956              Afterlife of the Party  MOVIE   
4915   ts255779                   Gabby's Dollhouse   SHOW   
5610   tm861018                        Sweet & Sour  MOVIE   
4411   tm839479                            Fan Girl  MOVIE   
4042   ts221041                     My Dear Warrior   SHOW   
4101   tm838765  A Christmas Prince: The Royal Baby  MOVIE   

                                            description  release_year  \
1729  After crash-landing on an alien planet, the Ro...          2018   
413   Dinosaur King is a card-based arcade game from...          2007   
3730  Wielding a light-up sword through the dark cor...          2020   
2844  Unbridled comic Chris D'Elia reconsiders his a...          2017   
5101  A social butterfly who dies during her birthda...          2021   
4915  This colorful series leads preschoolers room t...          2021   
5610  Faced with real-world opportunities and challe...          2021   
4411  A starstruck girl sneaks into an actor’s house...          2020   
4042  Sky is beautiful single female astronomer at a...          2019   
4101  Christmas brings the ultimate gift to Aldovia:...          2019   

     age_certification  runtime         genres production_countries  seasons  \
1729             TV-PG       49         action               ['US']      3.0   
413                NaN       20        western               ['JP']      2.0   
3730             TV-14       51         comedy               ['KR']      1.0   
2844               NaN       65  documentation               ['US']      NaN   
5101               NaN      109          drama               ['US']      NaN   
4915              TV-Y       24         comedy               ['US']      4.0   
5610               NaN      102         comedy               ['KR']      NaN   
4411               NaN      100          drama               ['PH']      NaN   
4042             TV-14       44        romance               ['TH']      1.0   
4101               NaN       84        romance               ['US']      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1729   tt5232792         7.3     98084.0           55.052       7.654  
413    tt1524993         5.6       737.0           28.909       8.200  
3730  tt12879522         6.4      1195.0           10.551       7.800  
2844   tt7068896         6.5      1729.0            3.004       6.200  
5101  tt11742798         5.8      8728.0           28.436       6.900  
4915   tt9165438         7.0       379.0           15.275       9.000  
5610  tt14599938         6.7      2508.0           57.384       7.800  
4411  tt11091664         6.6       382.0            2.273       7.600  
4042  tt11281912         4.9        21.0            2.737       9.300  
4101  tt10006006         5.4      6764.0           15.638       6.400

然后对`production_countries`也使用相同的流程进行拆分

In [10]:
cleaned_titles['production_countries'][1020]

"['IL', 'NO', 'ZA', 'AE', 'GB', 'IS', 'IE']"

In [11]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda x: eval(x))
cleaned_titles['production_countries'][1020]

['IL', 'NO', 'ZA', 'AE', 'GB', 'IS', 'IE']

In [12]:
cleaned_titles = cleaned_titles.explode('production_countries')
cleaned_titles.sample(10)

id                                   title   type  \
2929  tm447328  ReMastered: Who Killed Jam Master Jay?  MOVIE   
4943  tm981945         The Last Letter From Your Lover  MOVIE   
2423   ts75012                       Million Yen Women   SHOW   
4056  tm885669      The Christmas Chronicles: Part Two  MOVIE   
1615  tm316567                  HiGH&LOW: The Red Rain  MOVIE   
4839  tm456513                        Army of the Dead  MOVIE   
527   tm112235                   Kannathil Muthamittal  MOVIE   
4114  ts222803                      Maradona in Mexico   SHOW   
4937  tm413169             Peter Rabbit 2: The Runaway  MOVIE   
3412  tm465724                       See You Yesterday  MOVIE   

                                            description  release_year  \
2929  Run-D.M.C. DJ Jam Master Jay made a huge impac...          2018   
4943  A young journalist in London becomes obsessed ...          2021   
2423  Five beautiful but mysterious women move in wi...          2017   
4056  Kate Pierce is reluctantly spending Christmas ...          2020   
1615  The Amamiya Brothers, Masaki (Takahiro) and Hi...          2016   
4839  Following a zombie outbreak in Las Vegas, a gr...          2021   
527   A little girl is told by her parents that she ...          2002   
4114  The Dorados, Culiacan's local team, are at the...          2019   
4937  Peter Rabbit runs away from his human family w...          2021   
3412  As two teen prodigies try to master the art of...          2019   

     age_certification  runtime         genres production_countries  seasons  \
2929               NaN       58          music                   US      NaN   
4943               NaN      109          drama                   CA      NaN   
2423               NaN       30          drama                   JP      1.0   
4056                PG      112         family                   US      NaN   
1615               NaN      110         action                   JP      NaN   
4839                 R      148         action                   US      NaN   
527                NaN      137            war                   IN      NaN   
4114             TV-MA       33  documentation                   PE      1.0   
4937                PG       93      animation                   US      NaN   
3412               NaN       86       thriller                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
2929   tt9046560         6.5       786.0            1.705         6.5  
4943   tt1893273         6.6     17099.0           26.670         7.5  
2423   tt6686450         7.6      1835.0           10.133         6.8  
4056  tt11057644         6.0     29538.0           44.178         6.6  
1615   tt5923040         6.1       296.0            2.529         7.8  
4839   tt0993840         5.7    169348.0          131.869         6.3  
527    tt0312859         8.4      8491.0            3.739         7.4  
4114  tt11168104         7.3      1124.0            3.572         7.1  
4937   tt8376234         6.2     11233.0          130.318         7.2  
3412   tt8743064         5.2     10725.0            7.583         5.6

在处理完`cleaned_titles`的结构性问题后，查看`cleaned_credits`

In [13]:
cleaned_credits.sample(10)

person_id         id                name              character  \
42933    1030652   tm424162         Ching-i Pai                    Bat   
61932    1234688   tm839479           Gie Onida       Andeng's Husband   
65553    2058351   tm986168  Christiana Montoya           Abby (voice)   
65817     920942  tm1235785           Phil Wang                   Self   
55499      68715   tm884369           Noah Reid                   Self   
18838    1939973   tm158302       Suman Chowdry          Ajju's Mother   
46139     589304   tm447963       Brad Anderson                    NaN   
41455    1092312   tm445153  Aishwarya S. Menon                 Celina   
72920     600008   ts321254    Tahir Raj Bhasin  Vikrant Singh Chauhan   
47670    1171731   ts236531        Bruna Louise                    NaN   

           role  
42933     ACTOR  
61932     ACTOR  
65553     ACTOR  
65817     ACTOR  
55499     ACTOR  
18838     ACTOR  
46139  DIRECTOR  
41455     ACTOR  
72920     ACTOR  
47670     ACTOR

从抽样的10条数据来看，`cleaned_credits`符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，无需处理结构性问题

### 评估数据干净度

In [14]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

从info方法返回的概况信息来看，`cleaned_titles`共有17818条数据，其中`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量存在缺失值，将在后续进行评估和清理

还有，`release_year`表示年份，数据类型应为日期，还需要进行数据类型转换

In [15]:
cleaned_titles['release_year'] = pd.to_datetime(cleaned_titles['release_year'], format='%Y')
cleaned_titles['release_year']

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [16]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


从输出结果看，`cleaned_credits`共有77801个观察值，其中`character`变量存在缺失值

此外，`person_id`表示演职员id，数据类型应为字符串，要进行数据类型转换

In [17]:
cleaned_credits['person_id'] = cleaned_credits['person_id'].astype(str)
cleaned_credits['person_id']

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

#### 处理缺失数据

在`cleaned_titles`中，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量存在缺失值

由于影视作品的标题、描述、适龄认证、上映国家、电视剧季数、IMDB的ID、IMDB的投票数、TMDB的流行度、TMDB的评分，不影响我们根据IMDB评分找到各流派高评分作品的演员。所以这些变量的缺失值可以不用处理

那么还剩下`genres`、`imdb_score`变量的缺失值要处理，我们先查询出`genres`缺失的观察值

In [18]:
cleaned_titles.query('genres.isnull()')

id                                      title   type  \
212    tm255589                              One Last Shot  MOVIE   
619    tm341561                    Like Twenty Impossibles  MOVIE   
632     ts86241               Le Robe De Mariage Des Cieux   SHOW   
636    tm404676                       To and from New York  MOVIE   
637     tm89054                        Osuofia in London 2  MOVIE   
...         ...                                        ...    ...   
5799  tm1040959                 The Circle: The Afterparty  MOVIE   
5802   ts302434    Plastic Cup Boyz: Laughing My Mask Off!   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
212   In this low-budget short film, two best buddie...   1998-01-01   
619   Occupied Palestine: A serene landscape now poc...   2003-01-01   
632   It was with much difficulty that Ai Qing was a...   2004-01-01   
636   While covering a story in New York City, a Sea...   2006-01-01   
637   Osuofia return to his Nigerian village with a ...   2004-01-01   
...                                                 ...          ...   
5799  Stars of The Circle drop by to discuss Season ...   2021-01-01   
5802  Comedy collective The Plastic Cup Boyz pour ou...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime genres production_countries  seasons imdb_id  \
212                NaN       30    NaN                  NaN      NaN     NaN   
619                NaN       16    NaN                   PS      NaN     NaN   
632              TV-MA       63    NaN                  NaN      1.0     NaN   
636                NaN       82    NaN                   US      NaN     NaN   
637                NaN       72    NaN                   XX      NaN     NaN   
...                ...      ...    ...                  ...      ...     ...   
5799               NaN       35    NaN                   US      NaN     NaN   
5802               NaN       33    NaN                  NaN      1.0     NaN   
5840               NaN       26    NaN                  NaN      NaN     NaN   
5844               NaN       35    NaN                   US      NaN     NaN   
5848             PG-13       37    NaN                   US      NaN     NaN   

      imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
212          NaN         NaN            1.890         5.2  
619          NaN         NaN            0.812         6.5  
632          NaN         NaN            0.600         NaN  
636          NaN         NaN            1.401         5.8  
637          NaN         NaN            1.091         NaN  
...          ...         ...              ...         ...  
5799         NaN         NaN            1.882        10.0  
5802         NaN         NaN            0.683         NaN  
5840         NaN         NaN            1.179         7.0  
5844         NaN         NaN              NaN         NaN  
5848         NaN         NaN            1.296        10.0  

[63 rows x 15 columns]

由于`genres`是后续分析所必需的数据，我们将把这些缺失的观察值删除，并查看删除后该列的缺失值个数

In [19]:
cleaned_titles = cleaned_titles.dropna(subset=['genres'])
cleaned_titles['genres'].isnull().sum()

0

然后查询出`imdb_score`缺失的观察值

In [20]:
cleaned_titles.query('imdb_score.isnull()')

id                                title   type  \
0      ts300399  Five Came Back: The Reference Films   SHOW   
75     tm132164                 Bill Hicks: Sane Man  MOVIE   
145    ts251477                      My First Errand   SHOW   
145    ts251477                      My First Errand   SHOW   
145    ts251477                      My First Errand   SHOW   
...         ...                                  ...    ...   
5790  tm1094060               My Heroes Were Cowboys  MOVIE   
5791  tm1047429              Alan Saldaña: Locked Up  MOVIE   
5810  tm1225897                           Social Man  MOVIE   
5810  tm1225897                           Social Man  MOVIE   
5833   ts307884                           HQ Barbers   SHOW   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5790  Robin Wiltshire's painful childhood was rescue...   2021-01-01   
5791  Mexican comedian Alan Saldaña is back, poking ...   2021-01-01   
5810  Two competitive social media Influencers go he...   2021-01-01   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5790                PG       23  documentation                   US      NaN   
5791               NaN       49         comedy                  NaN      NaN   
5810               NaN       96         comedy                  NaN      NaN   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5790  tt15084326         NaN         NaN            3.145         7.7  
5791         NaN         NaN         NaN            6.670         6.0  
5810  tt20198164         NaN         NaN              NaN         NaN  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  

[785 rows x 15 columns]

由于`imdb_score`是后续分析所必需的数据，我们将把这些缺失的观察值删除，并查看删除后该列的缺失值个数

In [21]:
cleaned_titles = cleaned_titles.dropna(subset=['imdb_score'])
cleaned_titles['imdb_score'].isnull().sum()

0

然后再看cleaned_credits的数据，只有`character`变量存在缺失值

由于`character`变量表示角色姓名，与后续的数据分析无关，因此不处理这列存在缺失的观察值

#### 处理重复数据

在`cleaned_titles`中，同一部影视作品会存在多个流派和发行国家，而在上面的 步骤中已经将这些多个值的列扩展为多行。所以当每个变量都一样的观察值，才是重复数据

In [22]:
cleaned_titles.duplicated().sum()

0

输出结果为0，说明`cleaned_titles`没有重复数据

对于`cleaned_credits`，有的人可以既是演员也是导演，所以也是当所以变量都一样的观察值，才是重复数据

In [23]:
cleaned_credits.duplicated().sum()

0

输出结果为0，说明`cleaned_credits`也不存在重复数据

#### 处理不一致数据

针对`cleaned_titles`，不一致数据可能存在于`genres`和`character`变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。

In [24]:
cleaned_titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

从上面看出，`genres`列里并不存在不一致数据，各个值都在指代不同的流派

接下来，针对`production_countries`列也是一样的流程，利用`value_counts`方法，得到`production_countries`的列表里面各个值的出现次数

In [25]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

由于`value_counts`执行结果中有太多值，Pandas只会默认显示开头和结尾的一些值。要完整展示结果，可以把`display.max_rows`设置为`None`，即取消展示行数上限。

但因为我们只是在当前调用`value_counts`时才需要看完整结果，所以可以结合`option_context`，只更改临时上限。

In [26]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

从以上输出结果来看，出品国家都用两位的国家代码来表示，除了里面存在一个的`Lebanon`值。

`Lebanon`的国家代码是`LB`，出现了39次，说明此处数据不一致。`LB`和`Lebanon`都在表示同一国家，需要进行统一

把`cleaned_titles`里，`production_countries`的`"LB"`和`"Lebanon"`统一为`LB`，并检查替换后是否还存在`"Lebanon"`：

In [27]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].replace('Lebanon', 'LB')
len(cleaned_titles.query('production_countries == "Lebanon"'))

0

针对`cleaned_credits`，不一致数据可能存在于`role`中，我们将查看是否存在多个不同值指代同一演职员类型的情况

In [28]:
cleaned_credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

从以上输出结果来看，`role`只有两种可能的值，`ACTOR`或`DIRECTOR`，不存在不一致数据。我们可以把这列的类型转换为`Category`，好处是比字符串类型更节约内存空间，也能表明说值的类型有限

In [29]:
cleaned_credits['role'] = cleaned_credits['role'].astype('category')
cleaned_credits['role']

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### 处理无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [30]:
cleaned_titles.describe()

release_year       runtime      seasons    imdb_score  \
count                          16970  16970.000000  5954.000000  16970.000000   
mean   2015-11-14 22:42:51.974072064     80.912552     2.455492      6.514207   
min              1954-01-01 00:00:00      0.000000     1.000000      1.500000   
25%              2015-01-01 00:00:00     45.000000     1.000000      5.800000   
50%              2018-01-01 00:00:00     90.000000     2.000000      6.600000   
75%              2020-01-01 00:00:00    107.000000     3.000000      7.300000   
max              2022-01-01 00:00:00    225.000000    42.000000      9.500000   
std                              NaN     39.596172     2.869428      1.131095   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.694100e+04     16842.000000  16515.000000  
mean   3.281655e+04        29.396307      6.846933  
min    5.000000e+00         0.600000      1.000000  
25%    7.800000e+02         4.070000      6.200000  
50%    3.508000e+03        10.195000      6.900000  
75%    1.697800e+04        23.639000      7.500000  
max    2.294231e+06      2274.044000     10.000000  
std    1.141492e+05        93.178235      1.078831

从以上统计信息来看，`original_titles`里不存在脱离现实意义的数值。

`original_credits`由于不包含表示数值含义的变量，因此无需用`describe`检查。

## 整理数据

此次数据分析目标是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员

那为了能同时获得流派与演员数据，我们需要把`cleaned_titles`和`cleaned_credits`，通过`id`作为键进行连接，因为两个数据表中`id`都是影视作品ID

In [31]:
credits_with_titles = pd.merge(cleaned_credits, cleaned_titles, on="id", how="inner")
credits_with_titles.head()

person_id       id            name      character   role        title  \
0      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
1      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
2     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
3     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
4      7064  tm84618   Albert Brooks            Tom  ACTOR  Taxi Driver   

    type                                        description release_year  \
0  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  crime                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  crime                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.2    808582.0           40.965       8.179  
1         8.2    808582.0           40.965       8.179  
2         8.2    808582.0           40.965       8.179  
3         8.2    808582.0           40.965       8.179  
4         8.2    808582.0           40.965       8.179

由于我们只对挖掘演员的参演作品口碑感兴趣，导演不在我们的分析范围内，因此根据`role`，筛选出类型为`ACTOR`的观察值，供后续分析。

In [32]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

为了挖掘出各个流派中的高IMDB评分作品演员，我们需要先根据流派和演员进行分组。

对演员进行分组的时候，选择的是用`person_id`而不是`name`变量，原因是名字容易出现错拼或者重名的情况，演职员ID会比演员姓名更加准确地反映是哪位演员

In [33]:
groupby_genres_and_person_id = actor_with_titles.groupby(['genres', 'person_id'])

分组后，我们只需要对`imdb_score`的值进行聚合计算，因此只提取`imdb_score`变量，然后调用`mean`，来计算各个流派影视作品中，每位演员参演作品的平均IMDB评分

In [34]:
imdb_score_groupby_genres_and_person_id = groupby_genres_and_person_id['imdb_score'].mean()
imdb_score_groupby_genres_and_person_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

我们可以调用`reset_index`，对层次化索引进行重置，得到更加规整的DataFrame

In [35]:
imdb_score_groupby_genres_and_person_id_df = imdb_score_groupby_genres_and_person_id.reset_index()
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

现在针对流派和演员分组的IMDB评分数据已经整理好，可以进入后续的分析步骤了。

但我们当前可以继续做一些数据整理，比如对上面的结果再次进行分组，找出各个流派里演员作品最高的平均评分是多少、最高评分对应的演员名字是什么。

要得到这一结果，我们需要再次用`genres`进行分组，然后提取出`imdb_score`变量，计算其最大值

In [36]:
genres_max_scores = imdb_score_groupby_genres_and_person_id_df.groupby('genres')['imdb_score'].max()
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

在我们知道最高分后，可以把以上结果和之前得到的`imdb_score_groupby_genres_and_person_id_df`再次进行连接，得到最高分对应的各个演员ID是什么，也就是这个最高平均分是哪位演员拿到的

In [37]:
genres_max_score_with_person_id = pd.merge(imdb_score_groupby_genres_and_person_id_df, genres_max_scores, on=['genres', 'imdb_score'])
genres_max_score_with_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

从以上结果可以看出，最高分对应的演员不一定只有一位，可能有多位演员的平均得分相同。

为了得到演员ID所对应的演员名字，我们可以和`cleaned_credits`这个DataFrame进行连接。这个DataFrame还有其它列，我们只需要得到`person_id`和`name`的对应，所以可以先提取出那两列，并把重复行删除

In [38]:
actor_id_with_names = cleaned_credits[['person_id', 'name']].drop_duplicates()
actor_id_with_names.head(10)

person_id             name
0      3748   Robert De Niro
1     14658     Jodie Foster
2      7064    Albert Brooks
3      3739    Harvey Keitel
4     48933  Cybill Shepherd
5     32267      Peter Boyle
6    519612   Leonard Harris
7     29068   Diahnne Abbott
8    519613      Gino Ardito
9      3308  Martin Scorsese

下一步就可以把`actor_id_with_names`与前面得到的`genres_max_score_with_person_id`进行连接，增加`name`变量，从而展示平均评分最高的演员名字

In [39]:
genres_max_score_with_actor_name = pd.merge(genres_max_score_with_person_id, actor_id_with_names, on="person_id")
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0       action     12790         9.3         Olivia Hack
1        scifi     12790         9.3         Olivia Hack
2       action      1303         9.3       Jessie Flower
3    animation      1303         9.3       Jessie Flower
4       family      1303         9.3       Jessie Flower
..         ...       ...         ...                 ...
131        war    826547         8.8         Yuto Uemura
132    western     22311         8.9     Koichi Yamadera
133    western     28166         8.9  Megumi Hayashibara
134    western     28180         8.9      Unsho Ishizuka
135    western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

为了把相同流派都排序在一起，我们还可以用`sort_values`方法，把结果里面的行根据`genres`进行排序，然后用`reset_index`把索引重新排序。

索引重新排序后，DataFrame会多出`index`一列，我们可以再把`index`列进行删除

In [40]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action    336830         9.3     André Sogliuzzo
2     action     21033         9.3          Zach Tyler
3     action     86591         9.3       Cricket Leigh
4     action      1303         9.3       Jessie Flower
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     28166         8.9  Megumi Hayashibara
133  western     28180         8.9      Unsho Ishizuka
134  western     22311         8.9     Koichi Yamadera
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]